orchestrator/careplancontributor/service.go

```diff
@@ -117,6 +126,86 @@ func (s Service) handleProxyToEPD(writer http.ResponseWriter, request *http.Requ
+	// Authorize requester before proxying FHIR request
+	// Data holder must verify that the requester is part of the CareTeam by checking the URA
+	// Validate by retrieving the CarePlan from CPS, use URA in provided token to validate against CareTeam
+	// CarePlan should be provided in X-Scp-Context header
+	carePlanURLValue := request.Header[carePlanURLHeaderKey]
+	if len(carePlanURLValue) == 0 {
+		return errors.New(fmt.Sprintf("%s header value must be set", carePlanURLHeaderKey))
+	}
+	if len(carePlanURLValue) != 1 {
+		return errors.New(fmt.Sprintf("%s header must only contain one value", carePlanURLHeaderKey))
+	}
+	carePlanURL := carePlanURLValue[0]
+	if carePlanURL == "" {
+		return errors.New(fmt.Sprintf("%s header value must be set", carePlanURLHeaderKey))
+	}
+	if !strings.HasPrefix(carePlanURL, s.carePlanServiceURL.String()) {
+		return errors.New("invalid CarePlan URL in header")
+	}
+	u, err := url.Parse(carePlanURL)
+	if err != nil {
+		return err
+	}
+
+	carePlanServiceClient := fhirclient.New(s.carePlanServiceURL, s.scpHttpClient, coolfhir.Config())
+	var bundle fhir.Bundle
+	// Use extract CarePlan ID to be used for our query that will get the CarePlan and CareTeam in a bundle
+	carePlanId := strings.TrimPrefix(strings.TrimPrefix(u.Path, "/fhir/CarePlan/"), s.carePlanServiceURL.String())
+	newUrl := fmt.Sprintf("CarePlan?_id=%s&_include=CarePlan:care-team", carePlanId)
+	err = carePlanServiceClient.Read(newUrl, &bundle)
+	if err != nil {
+		return err
+	}
+	// We are expecting a total of 1 result, with 2 entries (one for CarePlan, one for CareTeam)
+	if *bundle.Total == 0 {
+		return errors.New("returned bundle has no results for CarePlan")
+	}
+	if len(bundle.Entry) < 2 {
+		return errors.New(fmt.Sprintf("returned bundle has incorrect number of entries %d expecting at least 2", len(bundle.Entry)))
+	}
+
+	var careTeams []fhir.CareTeam
+	err = coolfhir.ResourcesInBundle(&bundle, coolfhir.EntryIsOfType("CareTeam"), &careTeams)
+	if err != nil {
+		return err
+	}
+	if len(careTeams) == 0 {
+		return errors.New("CareTeam not found in bundle")
+	}
+
+	// Validate CareTeam participants against requester
+	principal, err := auth.PrincipalFromContext(request.Context())
+	if err != nil {
+		return err
+	}
+	isValidRequester := false
+	for _, careTeam := range careTeams {
+		for _, participant := range careTeam.Participant {
+			for _, identifier := range principal.Organization.Identifier {
+				if coolfhir.IdentifierEquals(participant.Member.Identifier, &identifier) {
+					// Member must have start date, this date must be in the past, and if there is an end date then it must be in the future
+					err = coolfhir.ValidateCareTeamParticipantPeriod(participant)
+					if err != nil {
+						return err
+					}
+					isValidRequester = true
+					break
+				}
+			}
+		}
+	}
+
+	if !isValidRequester {
+		return errors.New("requester does not have access to resource")
```

Comment:
Would be cool if we could return something like HTTP 403 in this case. Maybe with a custom error type that lets you specify the HTTP status code?

Reply:
That sounds like a good addition, will add it